File: studio/src/features/agents/status/runScopes.ts
import type { AgentRunScope, RunRecord } from "./types";

/**
 * The scope of a run with no provider at all, whose durable terminal session
 * hosts a plain interactive shell. It is the one scope whose runs carry a null
 * `agent`.
 */
export const SHELL_RUN_SCOPE: AgentRunScope = "shell";

/** Scopes whose runs legitimately carry no provider. */
const AGENTLESS_SCOPES: ReadonlySet<AgentRunScope> = new Set([SHELL_RUN_SCOPE]);

/**
 * Whether a run has no provider.
 *
 * Every surface that reads a run's `agent` needs an answer to this, and the
 * answer is a property of the run's scope rather than of the field happening to
 * be missing on the wire — so readers branch on the scope and never invent a
 * substitute provider slug.
 */
export function isAgentlessRun(run: Pick<RunRecord, "scope">): boolean {
  return AGENTLESS_SCOPES.has(run.scope);
}

File: backend/apps/runs/run_scopes.py
"""The durable routing scopes a run can carry.

A run's scope is its own discriminator — hooks can report lifecycle before the
terminal-session mirror exists, so the run row owns it. ``shell`` is the one
scope whose runs have no provider at all: a shell run's ``agent`` is null and
nothing about it may be interpreted as an agent run (#665).
"""

from __future__ import annotations

#: Every scope the persisted run record recognises, in wire order.
RUN_SCOPES: tuple[str, ...] = ("task", "plan", "instant", "docchat", "shell")

#: The scope of a run with no agent, hosting a plain interactive shell.
SHELL_SCOPE = "shell"

#: Scopes whose runs legitimately carry no ``agent``.
AGENTLESS_SCOPES = frozenset({SHELL_SCOPE})


def is_agentless_scope(scope: str | None) -> bool:
    """Whether a run in ``scope`` is expected to have no provider."""

    return scope in AGENTLESS_SCOPES
